develop-docs/engineering-practices/rust.mdx

```diff
@@ -50,26 +50,26 @@ During migration you may need normal functions which return futures, for their s
 
 ### Async Traits
 
-In **traits** you can not yet use `async fn` ([see this blog post](https://smallcultfollowing.com/babysteps/blog/2019/10/26/async-fn-in-traits-are-hard/)).
-In this case, functions should return `-> Pin<Box<dyn Future<Output = ...> + Send>>`:
+Support for async in **traits** has [landed in Rust](https://blog.rust-lang.org/2023/12/21/async-fn-rpit-in-traits.html)
+and should generally be preferred now.
 
 ```rust
-trait Database {
-    fn get_user(&self) -> Pin<Box<dyn Future<Output = User> + Send + '_>>;
+pub trait Database {
+    fn get_user(&self) -> impl Future<Output = User> + Send;
```

Comment:
Wanna add a callout here why we should be explicit with `Send` and `Sync` and prefer `impl Future` over `async` for that reason?

Personally, I'd add a _recommendation_ to name types where possible (especially for Iterators!) as this is extremely useful for composition in the absence of generators, but for `async` there's simply no need to create named futures in most cases.

Reply:
> Wanna add a callout here why we should be explicit with Send and Sync and prefer impl Future over async for that reason?

I kept this message, hoping it would address this clearly enough:

> Note that the returned future type is `Send`, to ensure that it can run on a multi-threaded runtime.

> Personally, I'd add a recommendation to name types where possible (especially for Iterators!) as this is extremely useful for composition in the absence of generators [...]

This requires explicit associated types until this restriction is lifted. I think for most of Sentry's (application) code this would be overkill. I can add another note that explains that as well.